### ISSUE TITLE
New operator: fallback

### ISSUE THREAD

**Issue by AdamSLevy** [prometheus/prometheus#9624]
## Proposal

In a perfect world we know exactly what recording rules we should add to our metrics systems from day 1. The reality is that recording rules get introduced over time as we realize that certain queries are useful and computationally expensive enough to justify a rule. 

After introducing a new rule, to maintain continuity in dashboards, it is tempting to reach for the `or` operator.

```
job:http_inprogress_requests:sum or sum by (job) (http_inprogress_requests)
```

Unfortunately when you are querying a time range when the recorded metric exists, it is much less efficient than just querying only for the recorded metric. In fact it is less efficient than just doing one or the other query individually for any time range, regardless of whether the recorded metric exists. The reason is that the promQL engine still has to perform the right hand side query and match up label sets to evaluate the 'or'. 

What I really want in this case is a `fallback` operator, which only evaluates the right hand side if exactly NO time series exist on the left hand side. Because the presence of any data on the left hand side indicates that the recording rule was present at that evaluation time, there is no need to continue to process the right hand side query, because I'm not interested in matching up the timeseries label sets. The recording rule is all or nothing. It's either present, in which case there is no need for the "or" behavior, or its missing, in which case, yes please go ahead and compute the right hand side query, as a _fallback_.

**Comment by AdamSLevy** [comment]
I'll add that backfilling, while theoretically possible is not an ideal work around for this especially if you are maintaining metrics for a lot of multi-tenant customers where actually performing the backfill would require more considerations such as computation and storage resources. Despite improved support for backfilling rule data, it also requires some engineering to achieve, especially if using other long term storage solutions such as Thanos. The goal of recording rules is typically to make queries more efficient going forward, but backfilling is expensive and not worth it in many cases. But still, continuity of metrics and dashboards is a very nice thing to have. When a recorded metric is not available, if you're taking the time to look, its probably worth computing it for that timeseries. Thus having this query would make it possible to maintain dashboard continuity.

**Comment by roidelapluie** [comment]
Thank you for your request. I understand it but it seems really difficult to implement: PromQL indeed is computing all the select at the beginning of the query and that's probably the expensive part. I also think that it is quite confusing for users to understand what it would mean.

let's gather @juliusv and @beorn7 opinion on this as well.

**Comment by AdamSLevy** [comment]
Thanks for the response. I could be totally off base, but I'd imagine it implemented something like this:

- run the left hand side query
- If the last timestamp of the returned data is less than the END requested timestamp, then run the right hand side query for the remaining time range

**Comment by juliusv** [comment]
My first gut reaction is also:

* While I get the use case, it does seem like a fairly niche feature request (at least I can't remember anyone requesting something similar before)
* The required changes to PromQL under the hood would likely be pretty complex in relation to what it gives us, but @codesome or others who have written more engine code recently would understand that part better.
* It's adding to the user-visible surface of PromQL, which is another cost in terms of learning and confusion for users if most people will not use it (correctly).

**Comment by beorn7** [comment]
Various thoughts here:

I also believe that what @AdamSLevy describes needs a lot of changes under the hood of the PromQL engine. (Running parts of an expression first to then decide how to run the rest of the expression is a very different flow from what the engine is doing right now.)

_However_ I do think we should consider changes of that kind eventually. This is what I more or less mean when I say "PromQL still needs a query planner". I'm not sure if that's the correct terminology, but we often run into cases where the engine is doing "obviously" needless work. Some of them are easy to address, some are pretty tricky. Often the engine would need to evaluate a branch to then eliminate other branches rather than fully evaluate them. And that's what the engine cannot do at all right now. It's just evaluating the full tree always. Sometimes, you can write queries in certain ways to minimize the damage, but that's not always possible (and not always easy).

Therefore, it would be really nice to eventually get a proper query planner (again with the disclaimer that I might use the wrong technical term here). _However_ I do not think we should use that feature to introduce a fairly obscure `fallback` operator. What we should do is to make it all work automatically. Look at this query:
```PromQL
job:http_inprogress_requests:sum{job="example"} or sum by (job) (http_inprogress_requests{job="example"})
```
In this case, the (future version of) the engine can run the left side, and whenever there is a result at all, it can stop. But even with the more general query as originally quoted by @AdamSLevy, i.e.
```
job:http_inprogress_requests:sum or sum by (job) (http_inprogress_requests)
```
the engine could run the left side and then check out if there are any other `job` labels on any `http_inprogress_requests` series at all, and then only start the `sum` for those. I would expect that index lookup to be relatively inexpensive, while the `sum` aggregation is the actually expensive part (which the engine is currently fully doing, only to throw away most or even all of the results later).

In short: I'd love to see a query planner for PromQL, and once we have one, we can get most of the savings implicitly, without introducing a new operator.

**Comment by AdamSLevy** [comment]
Thanks everyone for the comments. I agree that if this could be addressed by improving performance of promQL with a query planner that would completely eliminate the need for this feature. In other words, if using `or` was as efficient (or close to) as just making the individual queries for the appropriate time ranges, then this language feature would not be necessary. 

For my application to work around this, I'm expanding our application config to include an `availableAfter` field which lets us know after what timestamp a recording rule is available to use. Any queries before the timestamp are made with the actual expression. Any queries after that timestamp can use the recording rule. The tricky bit is just the logic to split up a query that spans both before and after into two queries and stitching the results back together. More logic than I would have like to write but it actually ends up being more performant than any 'or' query.

Hope we get a query planner soon! Is that already on the roadmap/being tracked anywhere that I can watch?

I'll let someone else close this as I don't exactly know the policies for issues in this repo but feel free to.

**Comment by beorn7** [comment]
In my experience, the best thing to get such a more involved feature going is to discuss it at the monthly dev summit. I have added a proposal to the agenda already.

In most cases these days, it's not as hard to reach general consensus for the feature. However, someone has to create a design doc then, and eventually the work has to be invested to implement it. The many TODOs in our [list of design docs](https://prometheus.io/docs/introduction/design-doc/) speak volumes about all the heroic plans we have. But at least this page gives you an idea where domain experts would be very welcome if the came to the rescue.

There is also a nominal [roadmap](https://prometheus.io/docs/introduction/roadmap/) on the project site but I think it has stopped to represent the more ambitious plans of the project many years ago.

**Comment by bwplotka** [comment]
I would love to see some kind of planner of this kind. Similarly when you do `count(min_over_time(xxx[2w])` we could immediately skip all samples and start calculating a number of selected series with samples within a 2w period.

I think that is the natural evolution of PromQL engine, especially needed for efforts like https://github.com/thanos-io/thanos/issues/305

**Comment by bboreham** [comment]
Reviewed at the bug-scrub meeting.

The general concept of a query-planner was discussed at the dev summit; we are open to it.

We do not want a new operator.  The query optimizer should be smart enough to figure out what work is required.
Many people in the Prometheus community are interested in query optimization; for instance both Thanos and Mimir have query engine projects.

Since this issue is focused on the 'fallback' operator we will close it.
